### PR TITLE
extension of loading config file from general file system

### PR DESCRIPTION
**Pull request by autumnust**
- Replace the original config loading method with new one, which build by general file system api provided by hadoop.fs package. 

--- 

TODO: 
- `JobConfigurationManager` requires modification. What about testing ? [June 27:Addressed in #1066]

### CODE REVIEW

File: gobblin-api/src/main/java/gobblin/configuration/ConfigurationKeys.java [hunk 1]
```diff
@@ -59,6 +59,9 @@
   public static final long DEFAULT_JOB_CONFIG_FILE_MONITOR_POLLING_INTERVAL = 300000;
   // Directory where all job configuration files are stored
   public static final String JOB_CONFIG_FILE_DIR_KEY = "jobconf.dir";
+  // todo : Naming issue for this to be discussed later.
+  // Path where all job configuration files stored
+  public static final String JOB_CONFIG_FILE_GENERAL_PATH_KEY = "jobconf.path" ;
```
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/configuration/ConfigurationKeys.java]: Maybe use `"jobconf.fullyQualifiedPath"`?

File: gobblin-api/src/main/java/gobblin/configuration/ConfigurationKeys.java [hunk 2]
```diff
@@ -59,6 +59,9 @@
   public static final long DEFAULT_JOB_CONFIG_FILE_MONITOR_POLLING_INTERVAL = 300000;
   // Directory where all job configuration files are stored
```
Comment (ibuenros) [inline gobblin-api/src/main/java/gobblin/configuration/ConfigurationKeys.java]: Please modify this comment to specify this is assumed to be in local file system.

File: gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java [hunk 3]
```diff
@@ -140,12 +140,15 @@ public JobScheduler(Properties properties) throws Exception {
...
 
   @Override
+  // TODO: Testing the general configured Jobs routine.
   protected void startUp() throws Exception {
     LOG.info("Starting the job scheduler");
 
     this.scheduler.start();
     if (this.properties.containsKey(ConfigurationKeys.JOB_CONFIG_FILE_DIR_KEY)) {
-      scheduleLocallyConfiguredJobs();
+//      scheduleLocallyConfiguredJobs();
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java]: Please remove commented lines of code.

File: gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java [hunk 4]
```diff
@@ -367,6 +381,24 @@ private void scheduleLocallyConfiguredJobs() throws ConfigurationException, JobE
   }
 
   /**
+   * Load job configuration file(s) from general source
+   */
+  private List<Properties> loadGeneralJobConfigs() throws ConfigurationException, IOException {
+    List<Properties> jobConfigs = SchedulerUtils.loadGenericJobConfigs(this.properties);
+    LOG.info(String.format(jobConfigs.size() <= 1 ? "Loaded %d job configuration" : "Loaded %d job configurations",
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java]: Both sides of this conditional are the same.

File: gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java [hunk 5]
```diff
@@ -367,6 +381,24 @@ private void scheduleLocallyConfiguredJobs() throws ConfigurationException, JobE
...
+        jobConfigs.size()));
+    return jobConfigs ;
+  }
+
+  /**
+   * TODO: Support general file system API for configure file monitor.
+   * Job configuration file monitor using generic file system API.
+   */
+  private void startGeneralJobConfigFIleMonitor() throws Exception {
+    return ;
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java]: For void methods, omit the `return` statement. Also, consider using `UnsupportedOperationException`.

File: gobblin-runtime/src/test/java/gobblin/runtime/local/LocalJobLauncherTest.java [hunk 6]
```diff
@@ -23,7 +23,7 @@
 import gobblin.configuration.ConfigurationKeys;
 import gobblin.metastore.FsStateStore;
 import gobblin.metastore.StateStore;
-import gobblin.metastore.testing.ITestMetastoreDatabase;
+import gobblin.metastored.testing.ITestMetastoreDatabase;
```
Comment (ibuenros) [inline gobblin-runtime/src/test/java/gobblin/runtime/local/LocalJobLauncherTest.java]: Why this change?

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 7]
```diff
@@ -12,12 +12,16 @@
 
 package gobblin.util;
 
+import com.google.common.base.Charsets;
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: `com` imports should be listed at the end.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 8]
```diff
@@ -68,6 +76,35 @@ public boolean accept(File dir, String name) {
     }
   };
 
+  private static final PathFilter PROPERTIES_PATH_FILTER =        new PathFilter() {
+    @Override
+    public boolean accept(Path path) {
+      String fileExtension = path.getName().substring(path.getName().lastIndexOf('.') + 1);
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: You can use `path.getName().endsWith(JOB_PROPS_FILE_EXTENSION)` instead.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 9]
```diff
@@ -68,6 +76,35 @@ public boolean accept(File dir, String name) {
...
+      return fileExtension.equalsIgnoreCase(JOB_PROPS_FILE_EXTENSION);
+    }
+  };
+
+  private static final PathFilter NON_PROPERTIES_PATH_FILTER = new PathFilter() {
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: This is a negation of `PROPERTIES_PATH_FILTER`, it is generally preferable to explicitly make it a negation. e.g. `accept(Path)` method should be `!PROPERTIES_PATH_FILTER.accept(path)`.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 10]
```diff
@@ -68,6 +76,35 @@ public boolean accept(File dir, String name) {
...
+
+  /**
+   * Load job configuration from job configuration files stored in general file system,
+   * located by Path
+   */
+  public static List<Properties> loadGenericJobConfigs(Properties properties)
+      throws ConfigurationException, IOException {
+    List<Properties> jobConfigs = Lists.newArrayList();
+    loadGenericJobConfigsRecursive(jobConfigs, properties, getJobConfigurationFileExtensions(properties),
+        (new Path(properties.getProperty(ConfigurationKeys.JOB_CONFIG_FILE_DIR_KEY))).toUri() );
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: There is no reason to convert the last argument to `URI`, just pass the `Path` itself.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 11]
```diff
@@ -178,6 +215,96 @@ public boolean accept(File file) {
...
+   * Recursively load job configuration files under given URI of **directory of config files folder**
+   * todo
+   */
+  private static void loadGenericJobConfigsRecursive(List<Properties> jobConfigs, Properties rootProps,
+      Set<String> jobConfigFileExtensions, URI uri)
+      throws ConfigurationException, IOException {
+
+    /* create the generic file system */
+    org.apache.hadoop.conf.Configuration conf = new org.apache.hadoop.conf.Configuration();
+    FileSystem filesystem = FileSystem.get(uri, conf);
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: If the last argument of the method signature is of type `Path`, you can instead do `path.getFileSystem(conf)`.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 12]
```diff
@@ -178,6 +215,96 @@ public boolean accept(File file) {
...
   /**
+   * Recursively load job configuration files under given URI of **directory of config files folder**
+   * todo
+   */
+  private static void loadGenericJobConfigsRecursive(List<Properties> jobConfigs, Properties rootProps,
+      Set<String> jobConfigFileExtensions, URI uri)
+      throws ConfigurationException, IOException {
+
+    /* create the generic file system */
+    org.apache.hadoop.conf.Configuration conf = new org.apache.hadoop.conf.Configuration();
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: Is there a reason for the type to be fully qualified?

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 13]
```diff
@@ -178,6 +215,96 @@ public boolean accept(File file) {
...
+    FileSystem filesystem = FileSystem.get(uri, conf);
+
+    Path configDirPath = new Path(uri);
+    if (!filesystem.exists(configDirPath)) {
+      throw new RuntimeException("No specified .properties directory found :" + configDirPath.toString());
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: Why is this a .properties directory?

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 14]
```diff
@@ -178,6 +215,96 @@ public boolean accept(File file) {
...
+    }
+
+    FileStatus[] propertiesFilesStatus = filesystem.listStatus(configDirPath, PROPERTIES_PATH_FILTER);
+    ArrayList<String> propertiesFiles = new ArrayList<>();
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: There is no reason to use a `List` here. You could just do `filesystem.open(propertiesFilesStatus[0].getPath())` (after checking for uniqueness).

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 15]
```diff
@@ -178,6 +215,96 @@ public boolean accept(File file) {
...
+
+      // Load the properties, which may overwrite the same properties defined in the parent or ancestor directories.
+      // Open the inputStream, construct a reader and send to the loader for constructing propertiesConfiguration.
+      PropertiesConfiguration propertiesConfiguration = new PropertiesConfiguration();
+      Path uniqueConfigFilePath = configDirPath.suffix(propertiesFiles.get(0));
+      InputStreamReader inputStreamReader = new InputStreamReader(filesystem.open(uniqueConfigFilePath), Charsets.UTF_8);
+      propertiesConfiguration.load(inputStreamReader);
+
+      rootProps.putAll(ConfigurationConverter.getProperties(propertiesConfiguration));
+      inputStreamReader.close();
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: Use a try-with-resource block instead. i.e.
```java
try(InputStreamReader inputStreamReader = new InputStreamReader(filesystem.open(uniqueConfigFilePath), Charsets.UTF_8)) {
  propertiesConfiguration.load(inputStreamReader);
  rootProps.putAll(ConfigurationConverter.getProperties(propertiesConfiguration));
}
```

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 16]
```diff
@@ -178,6 +215,96 @@ public boolean accept(File file) {
...
+    }
+
+    // Get all non-properties files
+    FileStatus[] nonPropFiles = filesystem.listStatus(configDirPath, NON_PROPERTIES_PATH_FILTER);
+    ArrayList<String> names = new ArrayList<>();
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: Same comment as above regarding the unnecessary `List`

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 17]
```diff
@@ -178,6 +215,96 @@ public boolean accept(File file) {
...
+    for (FileStatus nonPropFile : nonPropFiles) {
+      names.add(nonPropFile.getPath().getName());
+    }
+    if (names.size() == 0 ) {
+      return;
+    }
+
+    for (String name : names) {
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: This should be `for (FileStatus fileStatus : nonPropFiles)` instead, avoiding an extra call to `getFileStatus`.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 18]
```diff
@@ -178,6 +215,96 @@ public boolean accept(File file) {
...
+        rootPropsCopy.putAll(rootProps);
+        loadGenericJobConfigsRecursive(jobConfigs, rootPropsCopy, jobConfigFileExtensions, uri);
+      } else {
+        if (!jobConfigFileExtensions.contains(
+            configFilePath.getName().substring(configFilePath.getName().lastIndexOf('.') + 1).toLowerCase())) {
+          LOGGER.warn("Skipped file " + configFilePath + " that has an unsupported extension");
+          continue;
+        }
+
+
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: Please remove extra empty lines.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 19]
```diff
@@ -178,6 +215,96 @@ public boolean accept(File file) {
...
+        // Then load the job configuration properties defined in the job configuration file
+        PropertiesConfiguration propertiesConfiguration = new PropertiesConfiguration();
+        InputStreamReader inputStreamReader = new InputStreamReader(filesystem.open(configFilePath), Charsets.UTF_8);
+        propertiesConfiguration.load(inputStreamReader);
+
+        jobProps.putAll(ConfigurationConverter.getProperties(propertiesConfiguration));
+        jobProps.setProperty(ConfigurationKeys.JOB_CONFIG_FILE_PATH_KEY, uri.toString() );
+        jobConfigs.add(jobProps);
+
+        inputStreamReader.close();
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: Same as above regarding try with resource.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 20]
```diff
@@ -178,6 +215,96 @@ public boolean accept(File file) {
...
+      }
+    }
+
+    filesystem.close();
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: Use try with resource for `filesystem`.

File: gobblin-utility/src/test/java/gobblin/util/SchedulerUtilsTest.java [hunk 21]
```diff
@@ -12,29 +12,31 @@
 
 package gobblin.util;
 
+import com.google.common.collect.Sets;
```
Comment (ibuenros) [inline gobblin-utility/src/test/java/gobblin/util/SchedulerUtilsTest.java]: Check expected order of imports here: http://gobblin.readthedocs.io/en/latest/developer-guide/CodingStyle/

File: gobblin-runtime/src/test/java/gobblin/runtime/local/LocalJobLauncherTest.java (hunk 6 above)
Comment (autumnust) [inline gobblin-runtime/src/test/java/gobblin/runtime/local/LocalJobLauncherTest.java]: A typo which caused the building error.

File: gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java (hunk 5 above)
Comment (autumnust) [inline gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java]: This is the interface for config monitor, will remove it temporarily.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java (hunk 10 above)
Comment (autumnust) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: The reason why I passed the URI is related to the implementation of `loadGenericJobConfigsRecursive`, in which there will be a line : `FileSystem filesystem = FileSystem.get(uri, conf); `  

Since FIleSystem constructor doesn't accepet Path, I introduce URI in.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java (hunk 13 above)
Comment (autumnust) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: There's a constant String named `JOB_PROPS_FILE_EXTENSION` with the value of "properties", that's why I put it so. 

Do we need to use ".pull" / ".job" ? I guess there's some historical problem which caused inconsistency.

File: gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanGobblinDaemon.java [hunk 22]
```diff
@@ -0,0 +1,31 @@
+package gobblin.azkaban;
+
+import java.util.Properties;
+import org.apache.log4j.Logger;
+import gobblin.scheduler.SchedulerDaemon;
+import azkaban.jobExecutor.AbstractJob;
+
+
+public class AzkabanGobblinDaemon extends AbstractJob {
```
Comment (ibuenros) [inline gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanGobblinDaemon.java]: Please add a javadoc on this class.

File: gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java [hunk 23]
```diff
@@ -356,6 +356,17 @@ private void scheduleLocallyConfiguredJobs() throws ConfigurationException, JobE
...
+    for (Properties jobProps : loadGeneralJobConfigs()) {
+      boolean runOnce = Boolean.valueOf(jobProps.getProperty(ConfigurationKeys.JOB_RUN_ONCE_KEY, "false"));
+      scheduleJob(jobProps, runOnce ? new RunOnceJobListener() : new EmailNotificationJobListener());
+    }
+  }
+
+  /**
    * Load local job configurations.
    */
   private List<Properties> loadLocalJobConfigs() throws ConfigurationException {
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java]: Please remove this class if it is unused.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 24]
```diff
@@ -12,32 +12,37 @@
...
-
+import com.google.common.base.Charsets;
 import com.google.common.base.Function;
 import com.google.common.base.Preconditions;
 import com.google.common.base.Splitter;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Iterables;
 import com.google.common.collect.Lists;
 import com.google.common.io.Files;
+import gobblin.configuration.ConfigurationKeys;
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: Please fix the format of the imports. There's an empty line missing before `com.google.common.base.Charsets`, and another one before `gobblin.configuration...`

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 25]
```diff
@@ -178,6 +210,81 @@ public boolean accept(File file) {
...
+   */
+  private static void loadGenericJobConfigsRecursive(List<Properties> jobConfigs, Properties rootProps,
+      Set<String> jobConfigFileExtensions, Path configDirPath)
+      throws ConfigurationException, IOException {
+
+    /* create the generic file system */
+    Configuration conf = new Configuration();
+    try (FileSystem filesystem = configDirPath.getFileSystem(conf)) {
+      if (!filesystem.exists(configDirPath)) {
+        throw new RuntimeException("No specified .properties directory found :" + configDirPath.toString());
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: Can you change the log message to something like `"The specified job configurations directory was not found: "`

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 26]
```diff
@@ -178,6 +210,81 @@ public boolean accept(File file) {
...
+      }
+
+      FileStatus[] propertiesFilesStatus = filesystem.listStatus(configDirPath, PROPERTIES_PATH_FILTER);
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: When you run `fileSystem.listStatus`, always check that the result is not null. The contract of this method allows it to return null in certain situations, usually throwing `NullPointerExceptions` afterwards.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java [hunk 27]
```diff
@@ -178,6 +210,81 @@ public boolean accept(File file) {
...
+          }
+        }
+      }
+    }
+  }
+
+  /**
    * Recursively load job configuration files under the given directory.
    */
   private static void loadJobConfigsRecursive(List<Properties> jobConfigs, Properties rootProps,
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: Please remove unused methods.

File: gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java [hunk 28]
```diff
@@ -145,7 +145,7 @@ protected void startUp() throws Exception {
 
     this.scheduler.start();
     if (this.properties.containsKey(ConfigurationKeys.JOB_CONFIG_FILE_DIR_KEY)) {
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java]: It seems you are checking for the presence of `jobconf.dir`, then using `jobconf.fullyQualifiedDir` to actually get the job conf location, and if `jobconf.dir` is not present, it won't run anything, am I right?
The correct logic should be:
* if `jobconf.fullyQualifiedDir` is present, run `scheduleGeneralConfiguredJobs`.
* else if `jobconf.dir` is present, then set `jobconf.fullyQualifiedDir` to `file:///<jobconf.dir>`
* else throw an error

File: gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java [hunk 29]
```diff
@@ -356,6 +356,17 @@ private void scheduleLocallyConfiguredJobs() throws ConfigurationException, JobE
   }
 
   /**
+   * Schedule Gobblin jobs in general position
+   */
+  private void scheduleGeneralConfiguredJobs() throws ConfigurationException, JobException, IOException {
+    LOG.info("Scheduling locally configured jobs");
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java]: These are not locally configured jobs.

File: gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java (hunk 27 above)
Comment (autumnust) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: Deleting this method will trigger some error on code involving monitoring. Or, we need to migrate  all code in monitor-dev into this. Then too much overlap between two PRs. 

Just added a deprecated annotation there and will finally remove this after merged with the monitor modification.

Is this process acceptable ?

Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: That's ok. When will it be removable?

Comment (autumnust) [inline gobblin-utility/src/main/java/gobblin/util/SchedulerUtils.java]: After first PR got merged, I can migrate related part in #1066 and create a new PR. That one will remove all the deprecate method. Since #1066 seems more not that trivial as replacing File with Path, I will focus on that then.

File: gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanGobblinDaemon.java [hunk 30]
```diff
@@ -0,0 +1,35 @@
+package gobblin.azkaban;
+
+import java.util.Properties;
+import org.apache.log4j.Logger;
+import gobblin.scheduler.SchedulerDaemon;
+import azkaban.jobExecutor.AbstractJob;
+
+
+/**
+ * Wrapper of {@link SchedulerDaemon}, specially used by Azekaban to launch a job scheduler daemon.
```
Comment (ibuenros) [inline gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanGobblinDaemon.java]: typo: Azekaban -> Azkaban

File: gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java [hunk 31]
```diff
@@ -116,45 +116,58 @@
...
   @Override
-  protected void startUp() throws Exception {
+  protected void startUp()
+      throws Exception {
     LOG.info("Starting the job scheduler");
 
     this.scheduler.start();
-    if (this.properties.containsKey(ConfigurationKeys.JOB_CONFIG_FILE_DIR_KEY)) {
-      scheduleLocallyConfiguredJobs();
+    if (this.properties.containsKey(ConfigurationKeys.JOB_CONFIG_FILE_GENERAL_PATH_KEY)) {
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java]: Switch the order of these statements (to reduce code repetition):
```java
if (this.properties.containsKey(ConfigurationKeys.JOB_CONFIG_FILE_DIR_KEY) &&
 !this.properties.containsKey(ConfigurationKeys.JOB_CONFIG_FILE_GENERAL_PATH_KEY)) {
this.properties.setProperty(ConfigurationKeys.JOB_CONFIG_FILE_GENERAL_PATH_KEY,
 +          "file://" + this.properties.getProperty(ConfigurationKeys.JOB_CONFIG_FILE_DIR_KEY));
}
scheduleGeneralConfiguredJobs();
startJobConfigFileMonitor();
```

**Comment by ibuenros** [comment]
Couple new comments, but it looks good.
After you address the comments, please pull the master branch `git pull --rebase upstream master`, squash your commits, and then we can merge.
